### ISSUE TITLE
Breaking change in jsonschema==3.1.0 for Python 2.7

### ISSUE THREAD

**Issue by logachev** [Julian/jsonschema#611]
Recent update `3.1.0` introduces a breaking change for Python 2.7.

Here is a code sample that works with `3.0.0` but is broken with `3.1.0` because of the switch from `re` to `js_regex`: https://github.com/Julian/jsonschema/pull/609

``` Python
from jsonschema import Draft4Validator as Validator
schema = {'type': 'string', 'pattern': r'\d+'}
validator = Validator(schema)
list(validator.iter_errors('123'))
```

Thrown exception: 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/jsonschema/validators.py", line 328, in iter_errors
    for error in errors:
  File "/usr/local/lib/python2.7/site-packages/jsonschema/_validators.py", line 200, in pattern
    not js_regex.compile(patrn).search(instance)
  File "/usr/local/lib/python2.7/site-packages/js_regex/_impl.py", line 39, in compile
    raise TypeError("pattern={!r} must be a unicode string".format(pattern))
TypeError: pattern='\\d+' must be a unicode string
```

**Comment by Julian** [comment]
Having a look...

**Comment by Julian** [comment]
CC @Zac-HD.

**Comment by Julian** [comment]
OK, I think I don't see how to fix this yet, seems likely that it needs to happen downstream in js-regex, but I at least see how this slipped by, which is the changes to `test_validators` here: https://github.com/Julian/jsonschema/commit/51fc2f7a90ea8609090ddfd2a45ca4724e594462#diff-523a4ebdb045b5a1613d6711add86c7f

Not sure how I missed that during the review, but pretty sure that modification to the tests is our culprit.

**Comment by Julian** [comment]
OK, given that there's 2 separate issues here (#612 and #611) and that this is fairly fundamental functionality, I've temporarily pushed out v3.1.1 reverting the switchover.

We'll add some tests for the two cases here (and anything else we missed), and then there likely will be a v3.1.2 that goes back again to js-regex again.

**Comment by Julian** [comment]
Looking back at the test changes, not really good that those were mixed in with a rather unrelated test case, so probably this should have been tested much more explicitly anyhow. Will need to add tests specifically here on passing bytes in.

**Comment by Zac-HD** [comment]
Just to check, do we want to support bytes on Python 3?  My understanding was that JS was always at least Unicode BMP, which would not fit in the bytes type...

**Comment by Julian** [comment]
I am fairly sure on Py3 the json module always returns Unicode yeah. But
needs double checking, I always forget.

**Comment by Julian** [comment]
@Zac-HD I think https://github.com/Julian/jsonschema/commit/204c0c6df35c9ba51aeba03e384435014204ba3d covers us for this one.

**Comment by logachev** [comment]
Thanks @Julian for a quick mitigation!

**Comment by Julian** [comment]
Don't fully thank me until we're re-released :)

(And then thank @Zac-HD for helping contribute something I had 0 desire to ever get into, so real props for that as soon as we get this out the door :)

**Comment by Zac-HD** [comment]
This is fixed in `js-regex == 1.0.1` :tada: 

(the metacharacters-in-sets thing is trickier, but in progress)
